# Biological-Data-Modeling
## Overview:
This repository contains two labs, focusing on biological data analysis and modelling using Python. 
## Lab 1: Inferring Protein Concentration from RNA Data:
ðŸ“Œ **Objective**: Analyze and model protein concentrations based on RNA expression data.
## Tasks:
+ Import biological dataset and clean missing data.
+ Perform exploratory data analysis (mean, variance, histograms).
+ Create a scatter-plots of RNA vs Protein concentrations for all cell cycle phases (S, G2, G1).
+ Fit a linear regression model to predict protein levels at different cell cycle phases.
+ Identify genes related to cell cycle and ribosomes.
+ Analyze GO term frequency and correlations.

## **Lab 2: Chemical Reactions Modeling**
ðŸ“Œ **Objective**: Simulate and analyze **chemical reaction dynamics**.
### **Tasks**
+ Derive **differential equations** for a reaction system.   
+ Solve ODEs using **`scipy.integrate.odeint`**.  
+ Plot **X and Y concentration changes** over time.
+ Implement **Gillespie algorithm** for stochastic simulations.
+ Compare **ODE model vs. Gillepsie model**.

## **Lab 3: Sequence Alignment and Hidden Markov Models (HMMs)**
ðŸ“Œ **Objective**: Implement sequence alignment algorithms and HMMs.
### **Tasks**
+ Create subsitution matrices such as BLOSUM50 for scoring amino acid replacements. 
+ Align protein sequences using **Needleman-Wunsch** and **Smith-Waterman** algorithms.  
+ Simulate sequences generated by **HMMs** with defined states and probabilities.
+ Apply the **Viterbi algorithm** to decode hidden states in DNA.
+ Analyze and visualize sequence alignment results and HMM state sequences w/ the appropriate plots.

## **Lab 4: Wright-Fisher Model and Markov Processes in Evolutionary Dynamics**
ðŸ“Œ **Objective**: Model genetic drift, selection, mutation and recombination processes in populations.
### **Tasks**
+ Simulate **Wright-Fisher** models with and without crossover, tracking mutant gene proportions.
+ Compute **Markov** transition matrices and analyze evolutionary dynamics over multiple generations.
+ Plot and compare results from **Wright-Fisher** simulations and **Markov processes*.
+ Interpret differences between asexual and sexual reproduction models.
+ Use **`matplotlib`** for visualizing genetic mutation frequencies and **population dynamics.**
